### PR TITLE
p2p: Serialize cmpctblock at most once in NewPoWValidBlock

### PR DESCRIPTION
**Pull request by maflcko**
Instead of serializing for each peer, serialize at most once and copy the raw data for each peer.

### CODE REVIEW

**Comment by theStack** [comment]
Concept ACK

IIRC I also tried to tackle this TODO some time ago but failed as move-semantics prevented to store the serialized message. This PR is now a good opportunity to learn about futures and `std::async`.

**Comment by MarcoFalke** [comment]
Yeah, it is not possible to unconditionally `std::move` the result of a `std::shared_future` here, since it is unknown if there is another call to the getter later on. (In theory it might be possible to avoid the last copy and do a move instead, but I am not sure if this is worth it, let alone easily possible).

**Review by shaavan — APPROVED** [review]
ACK fabe7843a767505abdc580581e330fb4f963c788

I was researching about this PR. I found [this](https://stackoverflow.com/a/46162820) useful article. It talks about how using `std::deferred` allows lazy calculations. If the value has been already calculated, it will not be recalculated, potentially decreasing the number of times serialization is done to one.

I have reviewed the code changes in this PR, and I found them agreeable.

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1591,7 +1593,9 @@ void PeerManagerImpl::NewPoWValidBlock(const CBlockIndex *pindex, const std::sha
 
             LogPrint(BCLog::NET, "%s sending header-and-ids %s to peer=%d\n", "PeerManager::NewPoWValidBlock",
                     hashBlock.ToString(), pnode->GetId());
-            m_connman.PushMessage(pnode, msgMaker.Make(NetMsgType::CMPCTBLOCK, *pcmpctblock));
+
+            const CSerializedNetMsg& ser_cpmptblock{lazy_ser.get()};
```
Comment (theStack) [inline src/net_processing.cpp:1597]: This is a typo I guess?
```suggestion
            const CSerializedNetMsg& ser_cmpctblock{lazy_ser.get()};
```

Comment (MarcoFalke) [reply src/net_processing.cpp]: Thx, fixed

**Comment by MarcoFalke** [comment]
Fixed typo

**Review by shaavan — APPROVED** [review]
reACK fa61dd44f99323d10b0122c13224bc5cdb5e3d2a

Changes since my last review:

- Fixed a typo. `ser_cpmptblock` -> `ser_cmpctblock`.
- Rebased over current master.

**Review by theStack — APPROVED** [review]
After doing some more research about the concepts of futures and lazy evaluation in C++, I'm convinced now that this change is correct. This seems to be the first time ever that `std::async` and `std::shared_future` are used in the Bitcoin Core codebase.

Code-review ACK fa61dd44f99323d10b0122c13224bc5cdb5e3d2a

**Review by jonatack — COMMENTED** [review]
Nice. Posthumous ACK.

**Comment by jnewbery** [comment]
Concept ACK.

Is there a reason you need a `std::shared_future` rather than a `std::future`? It never gets copied for usage by multiple threads, which I believe is the usage for a shared_future. Would this be equivalent to what you've used here:

```diff
-    const std::shared_future<CSerializedNetMsg> lazy_ser{
+    std::future<CSerializedNetMsg> lazy_ser{
```

Could the same future be used to replace the `most_recent_compact_block` global, so that we don't have to reserialize the compact block for peers that request the compact block in low bandwidth mode?

**Comment by MarcoFalke** [comment]
No, that'd be UB, see https://en.cppreference.com/w/cpp/thread/future/get

> `valid()` is false after a call to `std::future<T>::get()`
> The behavior is undefined if `valid()` is false before the call to `get()`

**Comment by MarcoFalke** [comment]
> Could the same future be used to replace the most_recent_compact_block global, so that we don't have to reserialize the compact block for peers that request the compact block in low bandwidth mode?

Likely, but I haven't checked.

**Comment by jnewbery** [comment]
> No, that'd be UB, see [en.cppreference.com/w/cpp/thread/future/get](https://en.cppreference.com/w/cpp/thread/future/get)

Ah! Thanks.

**Comment by MarcoFalke** [comment]
For reference, for anyone wondering: A copy from the lazy ser is 25% faster on my machine than serialize.

**Pull request merged**